nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/actions/AbstractBulkComputation.java

```diff
@@ -0,0 +1,154 @@
+        super(name, 1, 1);
+        this.documentIds = new ArrayList<>(size);
+        this.timer = timer;
+        this.size = size;
+    }
+
+    @Override
+    public void init(ComputationContext context) {
+        getLog().debug(String.format("Starting computation: %s, size: %d, timer: %dms", metadata.name(), size, timer));
+        context.setTimer("timer", System.currentTimeMillis() + timer);
+    }
+
+    @Override
+    public void processRecord(ComputationContext context, String inputStreamName, Record record) {
+        BulkBucket bucket = BulkCodecs.getBucketCodec().decode(record.getData());
+        String commandId = bucket.getCommandId();
+        if (currentCommandId == null) {
+            // first time we need to process something
+            loadCurrentBulkCommandContext(commandId);
+        } else if (!currentCommandId.equals(commandId)) {
+            // new bulk id computation - send remaining elements
+            processBatch(context);
+            documentIds.clear();
+            loadCurrentBulkCommandContext(commandId);
+        }
+        // process record
+        documentIds.addAll(bucket.getIds());
+        if (documentIds.size() >= size) {
+            processBatch(context);
+        }
+    }
+
+    @Override
+    public void processTimer(ComputationContext context, String key, long timestamp) {
+        processBatch(context);
+        context.setTimer("timer", System.currentTimeMillis() + timer);
+    }
+
+    @Override
+    public void destroy() {
+        getLog().debug(
+                String.format("Destroy computation: %s, pending entries: %d", metadata.name(), documentIds.size()));
+    }
+
+    protected Log getLog() {
+        return LogFactory.getLog(getClass());
+    }
+
+    protected void loadCurrentBulkCommandContext(String commandId) {
+        currentCommandId = commandId;
+        KeyValueStore kvStore = Framework.getService(KeyValueService.class).getKeyValueStore(BULK_KV_STORE_NAME);
+        currentCommand = BulkCodecs.getCommandCodec().decode(kvStore.get(commandId + COMMAND));
+    }
+
+    protected void processBatch(ComputationContext context) {
+        if (!documentIds.isEmpty()) {
+            TransactionHelper.runInTransaction(() -> {
+                try {
+                    LoginContext loginContext = Framework.loginAsUser(currentCommand.getUsername());
+                    String repository = currentCommand.getRepository();
+                    try (CloseableCoreSession session = CoreInstance.openCoreSession(repository)) {
+                        compute(session, documentIds, currentCommand.getParams());
+                    } finally {
+                        loginContext.logout();
+                    }
+                } catch (LoginException e) {
+                    throw new NuxeoException(e);
+                }
+            });
+            BulkCounter counter = new BulkCounter(currentCommandId, documentIds.size());
```

Comment:
@pierre-gautier had extracted this to a method so that child classes can decide if they want to count or not.